### ISSUE TITLE
Add protocol for initialization mutators

### ISSUE THREAD

**Issue by nilsvu** [sxs-collaboration/spectre#4761]
All those type aliases used by `InitializeItems` are a prime example for a compile-time interface that must be documented properly. It's super confusing to have a struct in some `Actions` namespace/directory that defines all those type aliases, and no way to understand why they're there, what they do, how they're used, etc. We can't duplicate this information in code comments in every initialization mutator (which is what's currently done). A quick fix is to link `InitializeItems` in the docs of every initialize. A more maintainable solution is adding protocol. That probably should have been part of https://github.com/sxs-collaboration/spectre/pull/4486.

**Comment by nilsvu** [comment]
Possible solutions discussed today:

- The type aliases actually implement two different protocols: (1) argument_tags/return_tags/apply define the mutator, and (2) simple_tags/compute_tags/... are collected by InitializeItems and then collected from all actions somewhere in the parallel infrastructure.
- Therefore, add a protocol for both and make initialization mutators conform to both.
- Once the protocols exists, the type aliases can maybe default to empty lists. This won't catch typos in the alias names, but at least the protocol defines the alias names and is linked right from the implementation. Having to write all these type aliases everywhere is excessively verbose. Alternatively, simplify the interface by removing the need for so many type aliases.
- We could also group the aliases in nested structs to make clear they implement two different protocols.

**Comment by wthrowe** [comment]
As I expressed a few times during the meeting, I am unhappy with the idea of adding default values.  We've generally been moving away from defaults.

**Comment by nilsvu** [comment]
I don't like defaults either, but I also find 7 required type aliases in every action excessive (currently only initialization mutators require them and actions use empty lists as default). These are 7 lists of tags that all do something slightly different. I think this is a big mental load on everyone who just wants to write a small action. Do you have another suggestion @wthrowe ?

Edit: As written above, I think with the protocol documenting the interface, empty lists as defaults are fine.

**Comment by wthrowe** [comment]
Could you please be a bit more careful in this discussion with using the word "action"?  You're referring to initialization mutators as actions and also talking about real actions, and it's making you difficult to follow.

Anyway, my objection has nothing to do with documentation.

My problem with defaults for type aliases is there is no way to check that you've defined the correct thing.  There is no "using override" like there is for functions.  If you get it wrong, you may get an error in the same initialization mutator, you may get an error in some distant part of the code, or you may even get no error until someone later changes some unrelated thing.  These also aren't things that tend to get unit tested well.

Moreover, if I'm actually trying to see what one of these structs does, if the alias is there and is an empty list I can find it and know I have the answer.  If it defaults to empty, I have to search the struct (and possibly any base classes) to make sure I haven't missed the definition somewhere.  (And editor search functions are again subject to typos.)

I don't know what the decision on defaulting lists in actions would have been had we really considered it, but it wasn't practical because those lists were added long after most actions were written and there would have been no way to find and fix (almost) all of them aside from waiting for a compiler error for each occurrence.  (This might have been viable if actions had had a protocol.)

My suggestion is just to not default them to anything.

**Comment by kidder** [comment]
I agree with @wthrowe .   I think it is much cleaner and easier to understand when protocols (or structs that should have protocols like actions) define the expected type lists and to explicitly define them in classes rather than have them magically defaulted to empty type lists (and Will's point about potential errors is a good one).

In fact, now that we have a single DataBox type (instead of the variants), and all initialization actions are converted to mutators, we could eliminate some (all?) of these type aliases from actions.

**Comment by nilsvu** [comment]
I agree with everything you say @wthrowe. However, this doesn't solve the issue that I find the initialization mutators quite hard to understand and excessively verbose at the moment. Your point about making clear if I'm talking about actions or initialization mutators contributes to this: We're adding the mental load to people to understand the difference between the two. Therefore, they should at least be easy to understand. Having 7 different lists of tags that all do slightly different things is confusing. In this issue I'm trying to come up with a solution to that.
